Aflores plymouth (#234)
osisoft/Edge-Data-Store-Docs: content/sds/indexes/indexes-outside-dotnet.md

---
uid: IndexesOutsideDotNet
---

# Indexes outside of .NET framework

The following examples are in Python and JavaScript.

## Simple indexes

> When the .NET `SdsTypeBuilder` is unavailable, indexes must be built manually.

Reviewer: ```suggestion
When the .NET `SdsTypeBuilder` is unavailable, you must build indexes manually.
```

> The following examples show SdsTypes defined in [Python](https://github.com/osisoft/sample-ocs-waveform-python) and [Java Script](https://github.com/osisoft/sample-ocs-waveform-nodejs). For samples in other languages, go to [AVEVA Data Hub code samples in GitHub](https://github.com/osisoft/OSI-Samples-ADH/blob/main/docs/SDS_WAVEFORM.md).

Reviewer: ```suggestion
The following examples show `SdsTypes` defined in [Python](https://github.com/osisoft/sample-ocs-waveform-python) and [Java Script](https://github.com/osisoft/sample-ocs-waveform-nodejs). For samples in other languages, go to [AVEVA Data Hub code samples in GitHub](https://github.com/osisoft/OSI-Samples-ADH/blob/main/docs/SDS_WAVEFORM.md).
```

> To build an SdsType representation of sample classes in Python and Java Script, see [Sample](#sample) below:

Reviewer: ```suggestion
To build an `SdsType` representation of sample classes in Python and Java Script, see [Sample](#sample) below:
```

##### [Python](#tab/tabid-1)

```python
      class State(Enum):
        Ok = 0
        Warning = 1
        Alarm = 2
    
      class Simple(object):
        Time = property(getTime, setTime)
        def getTime(self):
          return self.__time
        def setTime(self, time):
          self.__time = time
    
        State = property(getState, setState)
        def getState(self):
          return self.__state
        def setState(self, state):
          self.__state = state
    
        Measurement = property(getValue, setValue)
        def getValue(self):
          return self.__measurement
        def setValue(self, measurement):
          self.__measurement = measurement
```

##### [JavaScript](#tab/tabid-2)

```javascript
      var State =
      {
        Ok: 0,
        Warning: 1,
        Alarm: 2
      }
    
      var Simple = function () {
        this.Time = null;
        this.State = null;
        this.Value = null;
      }
```

***

### Sample

> The following code is used to build an SdsType representation of the sample class above.

Reviewer: ```suggestion
The following code is used to build an `SdsType ` representation of the sample class above.
```

        "Id": "Warning",
        "Value": State.Warning
      });
    
      var stateTypePropertyAlarm = new SdsObjects.SdsTypeProperty({
        "Id": "Alarm",
        "Value": State.Alarm
      });
    
      var stateType = new SdsObjects.SdsType({
        "Id": "State",
        "Name": "State",
        "SdsTypeCode": SdsObjects.SdsTypeCodeMap.Int32Enum,
        "Properties": [stateTypePropertyOk, stateTypePropertyWarning,
          stateTypePropertyAlarm, stateTypePropertyRed]
      });
    
      // Value property is a simple non-indexed, pre-defined type
      var valueProperty = new SdsObjects.SdsTypeProperty({
        "Id": "Value",
        "SdsType": new SdsObjects.SdsType({
          "Id": "doubleType",
          "SdsTypeCode": SdsObjects.SdsTypeCodeMap.Double
        })
      });
    
      // Create the Simple SdsType
      var simpleType = new SdsObjects.SdsType({
        "Id": "Simple",
        "Name": "Simple",
        "Description": "This is a simple Sds type",
        "SdsTypeCode": SdsObjects.SdsTypeCodeMap.Object,
        "Properties": [timeProperty, stateProperty, valueProperty]
      });
```

***

> The `Time` property is identified as the primary index by defining its SdsTypeProperty as follows:

Reviewer: ```suggestion
The `Time` property is identified as the primary index by defining its `SdsTypeProperty` as follows:
```
